### DISCUSSION TITLE
Colored box

### DISCUSSION THREAD

**Discussion by ztef** [vsg-dev/VulkanSceneGraph#870] [General]
I draw a BOX using Builder, it works. How to draw a solid red one ?

I am trying to use GeomInfo with color, and colors attributes without success.  Any Idea ?

**Comment by robertosfield** [comment]
The vsgbuilder example has code paths that set up colours:

https://github.com/vsg-dev/vsgExamples/blob/master/examples/utils/vsgbuilder/vsgbuilder.cpp

If you can't spot the problem in your own code please post it so we can review it and see if we can spot the problem.

**Reply by ztef** [reply]
First, Thank you for your support,   Following is the issue :

Here is a minimal code for drawing a red Cube :

#include <vsg/all.h>
#include <iostream>


/*



    VulkanSceneGraph Minimal Example : Draw a RED Cube.



*/


int main(int argc, char** argv)
{
    
   
    auto windowTraits = vsg::WindowTraits::create();
    windowTraits->windowTitle = "RED CUBE";

   
    // Use a Builder
    auto builder = vsg::Builder::create();
 
    
    
    // Geometry and State variables 
    vsg::GeometryInfo geomInfo;

        geomInfo.dx.set(1.0f, 0.0f, 0.0f);
        geomInfo.dy.set(0.0f, 1.0f, 0.0f);
        geomInfo.dz.set(0.0f, 0.0f, 1.0f);

    vsg::StateInfo stateInfo;

        bool floatColors = true;
        stateInfo.wireframe = false;
        stateInfo.lighting = true;
        stateInfo.two_sided = false;


    // Create the scene
    auto scene = vsg::Switch::create();

            vsg::dvec3 centre = {0.0, 0.0, 0.0};
            double radius = 1.0;
    
            radius = vsg::length(geomInfo.dx + geomInfo.dy + geomInfo.dz);
    
            stateInfo.instance_positions_vec3 = true;

            auto positions = vsg::vec3Array::create(1);
            geomInfo.positions = positions;
        
            auto colors = vsg::vec4Array::create(1);
      
            for (auto& c : *(colors))
            {          
                c.set(1.0f,0.0f,0.0f,1.0f);
            }

            geomInfo.colors = colors;
        
            scene->addChild(true, builder->createBox(geomInfo, stateInfo));
           
            



    // create the viewer and assign window(s) to it
    auto viewer = vsg::Viewer::create();

    auto window = vsg::Window::create(windowTraits);
    if (!window)
    {
        std::cout << "Could not create windows." << std::endl;
        return 1;
    }

    viewer->addWindow(window);

    vsg::ref_ptr<vsg::LookAt> lookAt;
  

    // set up the camera
    lookAt = vsg::LookAt::create(centre + vsg::dvec3(0.0, -radius * 3.5, 0.0), centre, vsg::dvec3(0.0, 0.0, 1.0));

    double nearFarRatio = 0.001;
    auto perspective = vsg::Perspective::create(30.0, static_cast<double>(window->extent2D().width) / static_cast<double>(window->extent2D().height), nearFarRatio * radius, radius * 10.0);

    auto camera = vsg::Camera::create(perspective, lookAt, vsg::ViewportState::create(window->extent2D()));

   
    // add close handler to respond the close window button and pressing escape
    viewer->addEventHandler(vsg::CloseHandler::create(viewer));

    viewer->addEventHandler(vsg::Trackball::create(camera));



    // OPTION 1 :    Creating a commandGraph directly...it WORKS (Draws a red Cube)

        //auto commandGraph = vsg::createCommandGraphForView(window, camera, scene);
        //viewer->assignRecordAndSubmitTaskAndPresentation({commandGraph});


    // OPTION 2 :     Creating a view, then a renderGRaph   (Doesnt Work :      Draws a BLACK CUBE)

         auto view = vsg::View::create(camera, scene);

         auto renderGraph = vsg::RenderGraph::create(window);
        renderGraph->addChild(view);

        auto commandGraph = vsg::CommandGraph::create(window);
        commandGraph->addChild(renderGraph);

        viewer->assignRecordAndSubmitTaskAndPresentation({commandGraph});






    viewer->compile();

    

    // rendering main loop
    while (viewer->advanceToNextFrame() )
    {
        // pass any events into EventHandlers assigned to the Viewer
        viewer->handleEvents();

        viewer->update();

        viewer->recordAndSubmit();

        viewer->present();      
    }

    return 0;
}

**Reply by ztef** [reply]
As you can see : (if uncomment the OPTION 1) :   

auto commandGraph = vsg::createCommandGraphForView(window, camera, scene);
viewer->assignRecordAndSubmitTaskAndPresentation({commandGraph});

using the scene amd the camera to create a commandGRaph, it works fine and draw a red Box perfectly.



But  the following (creating a view and a rendergraph first, the scene appears to loose any color related parameters and only draw a BLACK Cube : 

    auto view = vsg::View::create(camera, scene);

     auto renderGraph = vsg::RenderGraph::create(window);
    renderGraph->addChild(view);

    auto commandGraph = vsg::CommandGraph::create(window);
    commandGraph->addChild(renderGraph);

    viewer->assignRecordAndSubmitTaskAndPresentation({commandGraph});

**Reply by ztef** [reply]
![Screenshot from 2023-07-10 17-46-56](https://github.com/vsg-dev/VulkanSceneGraph/assets/1700294/f36826ef-cf40-4936-aee0-f3b48eb3062c)
![Screenshot from 2023-07-10 17-47-44](https://github.com/vsg-dev/VulkanSceneGraph/assets/1700294/278bf566-b2fb-4e85-9d8d-d8cbf300bd0c)

**Reply by ztef** [reply]
it is a bug ? or I am missing something ?

**Reply by ztef** [reply]
![Screenshot from 2023-07-10 18-16-34](https://github.com/vsg-dev/VulkanSceneGraph/assets/1700294/d491889e-c05f-42bb-a7f0-fdd9d5d11e7a)
The reason I need the scene, is because the application i am trying to code is a scene selector, where each scene can handle their own world(camera, perspective etc...)

**Comment by robertosfield** [comment]
Hi Esteban,

I think the most likely reason for the black cube is that has been created with lighting on so needs a light source to illuminate it, and the vsg::createCommandGraphForView(window, camera, scene) convenience function assigns this to the vsg::View for you, but when you create the vsg::View by yourself it's in it's default state without any cameras assigned.  

The vsg::createCommandGraphForView() calls vsg::createRenderGraphForView() function which has the assignment of the Light:

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/app/RenderGraph.cpp#L187

Assigning one or more vsg::Light to View or elsewhere in your scene graph will illuminate the scene.

Robert.

**Comment by ztef** [comment]
Hi Robert, 

You are a genius !! it works !!  Thank you very much !!

By the way, what a wonderful project, congratulations.
